src/GitVersion.Core/Core/GitPreparer.cs

```diff
@@ -316,11 +316,10 @@ private void CreateOrUpdateLocalBranchesFromRemoteTrackingOnes(string remoteName
             var branchName = remoteTrackingReferenceName.Substring(prefix.Length);
             var localCanonicalName = "refs/heads/" + branchName;
 
-            var referenceName = ReferenceName.Parse(localCanonicalName);
```

Comment:
I think this line is needed - we need to parse the localCanonicalName

Reply:
Is the `ReferenceName.Parse` doing any validation? Because we are not using `referenceName` anywhere

Reply:
Yes it does https://github.com/GitTools/GitVersion/blob/3626ce5126e1375549224193d039820d521d3f77/src/GitVersion.Core/Git/ReferenceName.cs#L29

Reply:
The following validation snippet from `ReferenceName.Parse` : `IsPrefixedBy(canonicalName, LocalBranchPrefix)`
will match the constructed string `"refs/heads/" + branchName;`

What do you think about changing the construction of the string to `ReferenceName.LocalBranchPrefix + branchName;` instead?

Reply:
Well, the problem here is actually using the `IComparable` interface for searching a matching `IReference`. Doing such a linear lookupleads to O(n) algorithmic complexity, and as we are already within a loop, this leads into an O(n^2) algorithm complexity.

We should instead use an data structure with logarithmic lookups directly in order to take this first algorithmic complexity down to a O(log(n)), and the whole algorithm to a O(n log(n)) algorithmic complexity.

Lucky for us, the `IReferenceCollection` class does provide such an access (`IReference? this[string name] { get; }`), but the key of that dictionnary seems to be a `CanonicalName`.

I've modified my code so we use a `ReferenceName` to query the `IReferenceCollection` structure.